### DISCUSSION TITLE
Protobuf for geojson points?

### DISCUSSION THREAD

**Discussion by dhruv4kubra** [visgl/deck.gl#8154] [Q&A]
Hi,
I am looking to see if i could use protobuf for performance improvement. I see thats a recommeded way in "Performance Optimization" but nothing concrete is showcased as example or documentation in using it.

If there are links or references further that i could refer to will be greatly appreciated.

-Drew

**Comment by Pessimistress** [comment]
Protobuf does not speed up the layer on the rendering side, but it can help download data faster and save memory.

You can supply an encoded protobuf buffer to `data` and write a custom accessor to read from it. Search for "non-iterable" in https://deck.gl/docs/developer-guide/performance#supply-binary-blobs-to-the-data-prop. Something along the lines of

```proto
syntax="proto3"

message MyMessage {
  double longitude = 1;
  double latitude = 2;
}
```

```ts
import { MyMessage } from './my-message-compiled';
import { Reader } from 'protobufjs/minimal';

async function loadData() {
  const resp = await fetch(...);
  const buffer = await resp.arrayBuffer();

  return {
    length: 100000,
    reader: new Reader(new Uint8Array(buffer))
  };
}

new ScatterplotLayer({
  data: loadData(),
  getPosition: (_, {index, reader}) => {
    if (index === 0) {
      data.reader.pos = 0;  // reset the reader to read from the beginning
    }
    const message = MyMessage.decodeDelimited(data.reader);
    return [message.longitude, message.latitude];
  },
});
```
